sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubBufferedProducerClient.cs

```diff
@@ -873,5 +1063,277 @@ public virtual Task FlushInternalAsync(CancellationToken cancellationToken = def
+            }
+        }
+
+        /// <summary>
+        ///   Performs the actions needed to stop processing events. the <see cref="EventHubBufferedProducerClient" /> to begin accepting events to be
+        ///   enqueued and published.  Should this method be called while processing is not active, no action is taken.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the stop operation.</param>
+        ///
+        /// <remarks>
+        ///   This method will modify class-level state and should be synchronized.  It is assumed that callers hold responsibility for
+        ///   ensuring synchronization concerns; this method should be invoked after any primitives have been acquired.  Callers are also
+        ///   assumed to own responsibility for any validation that the client is in the intended state before calling.
+        ///
+        ///   This method does not consider any events that have been enqueued and are in a pending state.  It is assumed
+        ///   that the caller has responsibility for invoking <see cref="FlushInternalAsync" /> or <see cref="ClearInternalAsync" />
+        ///   prior to this method if needed.
+        /// </remarks>
+        ///
+        private async Task StopPublishingAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            Logger.BufferedProducerBackgroundProcessingStop(Identifier, EventHubName);
+
+            var capturedExceptions = default(List<Exception>);
+
+            try
+            {
+                // If there is no task running for the background management process, there
+                // is nothing to stop.
+
+                if (_producerManagementTask == null)
+                {
+                    return;
+                }
+
+                // Request cancellation of the background processing.
+
+                _publishingCancellationSource?.Cancel();
+                _publishingCancellationSource?.Dispose();
+                _publishingCancellationSource = new CancellationTokenSource();
+
+                // Wait for the background tasks to complete.
+
+                try
+                {
+                    await _producerManagementTask.ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // This is expected; no action is needed.
+                }
+                catch (Exception ex)
+                {
+                    Logger.BufferedProducerBackgroundProcessingStopError(Identifier, EventHubName, ex.Message);
+                    (capturedExceptions ??= new List<Exception>()).Add(ex);
+                }
+
+                // Clean up all partition publishers.
+
+                foreach (var pair in _activePartitionPublishers)
```

Comment:
For this scenario, do we care if the dictionary is modified while we are enumerating or whether the collection is a point in time snapshot?

Reply:
Snapshot should be safe here.  The method should only be invoked when the class state semaphore is held and is only referenced by `CloseAsync`.    The `_isClosed` flag gets set before processing is stopped, so the `Enqueue` methods that create the entries will fail "Is this closed" checks.

Reply:
I only ask because enumerating the values is not a snapshot, but doing `.ToArray()` first would be, I believe.  But seems like it shouldn't matter either way.

Reply:
Fair enough.   My intent was to use the [concurrency guarantee](https://docs.microsoft.com/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getenumerator?view=net-5.0#remarks) to catch anything added after I started cleaning up  - but pragmatically, that shouldn't be possible without call patterns being violated.    So, yeah.... snapshot is safe, real-time is better, but it should be the same net result.